src/steamship/agents/functions_based.py

```diff
@@ -0,0 +1,127 @@
+from steamship.agents.utils import build_chat_history
+from steamship.data.tags.tag_constants import ChatTag, RoleTag, TagKind, TagValueKey
+from steamship.data.tags.tag_utils import get_tag
+from steamship.plugin.capabilities import (
+    ConversationSupport,
+    FunctionCallingSupport,
+    SystemPromptSupport,
+)
+
+
+class FunctionsBasedAgent(Agent):
+    """Selects actions for AgentService based on a set of Tools."""
+
+    PROMPT = """You are a helpful AI assistant.
+
+NOTE: Some functions return images, video, and audio files. These multimedia files will be represented in messages as
+UUIDs for Steamship Blocks. When responding directly to a user, you SHOULD print the Steamship Blocks for the images,
+video, or audio as follows: `Block(UUID for the block)`.
+
+Example response for a request that generated an image:
+Here is the image you requested: Block(288A2CA1-4753-4298-9716-53C1E42B726B).
+
+Only use the functions you have been provided with."""
+
+    def __init__(self, llm: SteamshipLLM, tools: List[Tool], **kwargs):
+        super().__init__(tools=tools, **kwargs)
+        self.llm = llm
+        self.capabilities = [
+            SystemPromptSupport(),
+            ConversationSupport(),
+            FunctionCallingSupport(tools=tools),
+        ]
+        self.tools_map = {tool.name: tool for tool in tools}
+
+    def default_system_message(self) -> Optional[str]:
+        return self.PROMPT
+
+    def next_action(self, context: AgentContext) -> Action:
+        # Build the Chat History that we'll provide as input to the action
+        messages = build_chat_history(self.default_system_message(), self.message_selector, context)
+        # get working history (completed actions)
+        messages.extend(self._function_calls_since_last_user_message(context))
+
+        # Run the default LLM on those messages
+        output_blocks = self.llm.generate(messages=messages, capabilities=self.capabilities)
+
+        for block in output_blocks:
+            if block.mime_type == MimeTypes.STEAMSHIP_PLUGIN_FUNCTION_CALL_INVOCATION:
+                invocation = FunctionCallingSupport.FunctionCallInvocation.from_block(block)
+                tool = self.tools_map.get(invocation.tool_name)
+                if tool is None:
+                    raise SteamshipError(
+                        f"LLM attempted to invoke tool {invocation.tool_name}, but {self.__class__.__name__} does not have a tool with that name."
+                    )
+                # TODO (PR): Could overload Action for this case to piggy back Invocations on it, but ehhhhhhhhh...
+                # TODO Block parse for input.  text is the default argument that we currently pass in for Tools.  As
+                #  part of a refactor to allow for other parameters, this would need to change.
+                future_action = Action(
+                    tool=tool.name, input=[Block(text=invocation.args.get("text"))], output=None
+                )
```

Comment:
This is a tad awkward at the moment without doing an args / tools refactor as well.